src/openms/source/ANALYSIS/ID/SimpleSearchEngineAlgorithm.cpp

```diff
@@ -0,0 +1,654 @@
+#ifdef _OPENMP
+#pragma omp critical (residuedb_access)
+#endif
+        {
+          AASequence aas = AASequence::fromString(current_peptide);
+          ModifiedPeptideGenerator::applyFixedModifications(fixed_modifications.begin(), fixed_modifications.end(), aas);
+          ModifiedPeptideGenerator::applyVariableModifications(variable_modifications.begin(), variable_modifications.end(), aas, modifications_max_variable_mods_per_peptide_, all_modified_peptides);
+        }
+
+        for (SignedSize mod_pep_idx = 0; mod_pep_idx < (SignedSize)all_modified_peptides.size(); ++mod_pep_idx)
+        {
+          const AASequence& candidate = all_modified_peptides[mod_pep_idx];
+          double current_peptide_mass = candidate.getMonoWeight();
+
+          // determine MS2 precursors that match to the current peptide mass
+          multimap<double, Size>::const_iterator low_it;
+          multimap<double, Size>::const_iterator up_it;
+
+          if (precursor_mass_tolerance_unit_ppm) // ppm
+          {
+            low_it = multimap_mass_2_scan_index.lower_bound(current_peptide_mass - 0.5 * current_peptide_mass * precursor_mass_tolerance_ * 1e-6);
+            up_it = multimap_mass_2_scan_index.upper_bound(current_peptide_mass + 0.5 * current_peptide_mass * precursor_mass_tolerance_ * 1e-6);
+          }
+          else // Dalton
+          {
+            low_it = multimap_mass_2_scan_index.lower_bound(current_peptide_mass - 0.5 * precursor_mass_tolerance_);
+            up_it = multimap_mass_2_scan_index.upper_bound(current_peptide_mass + 0.5 * precursor_mass_tolerance_);
+          }
+
+          // no matching precursor in data
+          if (low_it == up_it) { continue; }
+
+          // create theoretical spectrum
+          PeakSpectrum theo_spectrum;
+
+          // add peaks for b and y ions with charge 1
+          spectrum_generator.getSpectrum(theo_spectrum, candidate, 1, 1);
+
+          // sort by mz
+          theo_spectrum.sortByPosition();
+
+          for (; low_it != up_it; ++low_it)
+          {
+            const Size& scan_index = low_it->second;
+            const PeakSpectrum& exp_spectrum = spectra[scan_index];
+            // const int& charge = exp_spectrum.getPrecursors()[0].getCharge();
+            const double& score = HyperScore::compute(fragment_mass_tolerance_, fragment_mass_tolerance_unit_ppm, exp_spectrum, theo_spectrum);
+
+            if (score == 0) { continue; } // no hit?
+
+            // add peptide hit
+            AnnotatedHit_ ah;
+            ah.sequence = c;
+            ah.peptide_mod_index = mod_pep_idx;
+            ah.score = score;
+
+#ifdef _OPENMP
+            omp_set_lock(&(annotated_hits_lock[scan_index]));
+            {
+#endif
+              annotated_hits[scan_index].push_back(ah);
+
+              // prevent vector from growing indefinitly (memory) but don't shrink the vector every time
+              if (annotated_hits[scan_index].size() >= 2 * report_top_hits_)
```

Comment:
this might work well for large values of `report_top_hits_` but for the default value of 1, this will cause shrinking for every second hit:

- add hit, vector.size() = 1
- add hit, vector.size() = 1
- add hit, vector.size() = 2
- add hit, vector.size() = 3 -> shrink to 1
- add hit, vector.size() = 2
- add hit, vector.size() = 3 -> shrink to 1
...

Reply:
I once used a priority queue and it was slightly slower. One reason might be that no reallocation is happening and the only thing happening during the resize is calling the destructor of AnnotatedHit. Maybe this is a noop for this simple struct?